app/src/main/kotlin/compiler/Compiler.kt

```diff
@@ -40,17 +54,72 @@ class Compiler(val diagnostics: Diagnostics) {
 
             val programProperties = ProgramAnalyzer.analyzeProgram(astWithBuiltinFunctions, diagnostics)
 
-            val functionCFGs = createGraphForProgram(astWithBuiltinFunctions, programProperties, diagnostics, diagnostics.hasAnyError())
+            val functionDetailsGenerators = FunctionDependenciesAnalyzer.createFunctionDetailsGenerators(
+                astWithBuiltinFunctions,
+                programProperties.variableProperties,
+                programProperties.functionReturnedValueVariables,
+                diagnostics.hasAnyError()
+            )
+
+            val functionCFGs = createGraphForProgram(astWithBuiltinFunctions, programProperties, functionDetailsGenerators, diagnostics)
+
+            val mainFunction = (
+                ast.globals.find { it is Program.Global.FunctionDefinition && it.function.name == "główna" } as Program.Global.FunctionDefinition?
+                )?.function
+            if (mainFunction == null) {
+                diagnostics.report(Diagnostic.MainFunctionNotFound())
+            }
 
             if (diagnostics.hasAnyError())
                 return false
 
-            // TODO:
-            // - make sure "główna" is present and generate a jump to it from "main",
-            // - linearize CFG for each function,
-            // - run liveness analysis on each linear function code,
-            // - run register allocation on each linear function code,
-            // - write ASM to output (display, global variables, functions)
+            val linearFunctions = functionCFGs.mapValues { Linearization.linearize(it.value, covering) }
+
+            val finalCode = linearFunctions.mapValues {
+                Allocation.allocateRegistersWithSpillsHandling(
+                    it.value,
+                    Liveness.computeLiveness(it.value),
+                    Allocation.REGISTER_ORDER,
+                    ColoringAllocation
+                )
+            }
+
+            finalCode.entries.forEach { functionDetailsGenerators[it.key]!!.spilledRegistersOffset.settledValue = it.value.spilledOffset.toLong() }
+
+            output.println("SECTION .bss")
```

Comment:
I would suggest to move the code with the actual printing of section headers and the `main` label into to the `lowlevel` package. It could be a function (say, `printAsmFile`), which would take three functions as arguments, corresponding to the three sections, and call these functions after printing the headers of the corresponding sections. In this way we would have a readable template with the layout of output assembly file.